uni/lib/unittest.icn

```diff
@@ -0,0 +1,166 @@
+# TODO: assertions
+#     - asserting objects
+#     - asserting nested values (list, table, objects, etc.)
+# TODO: Test Reporter
+#     - add message to results
+#     - add backtrace if possible
+# TODO: exit 0 when success, 1 when failure.
+# TODO: there should be a way to discover and load tests dynamically
+# TODO: add timer, see tests/unicon/tester.icn
+# TODO: how to catch stdin/stdout/stderr within tests, see capture().
+# TODO: enable/disable verbosity.
+
+package unittest
+
+link ximage
+
+import lang
+
+invocable all
+
+class TestCase : Object(__status)
+   $define STATUS_SUCCESS 0
+   $define STATUS_FAIL 1
+
+   method setupClass()
+      write(classname(self), "::setupClass()")
+   end
+
+   method teardownClass()
+      write(classname(self), "::teardownClass()")
+   end
+
+   method setup()
+      write(classname(self), "::setup()")
+   end
+
+   method teardown()
+      write(classname(self), "::teardown()")
+   end
+
+   method __reset()
+      self.__status := STATUS_SUCCESS
+   end
+
+   method assert(expression)
+      if not @expression[1] then {
+         self.__status := STATUS_FAIL
+         write("Assertion failed.")
+         fail
+      }
+
+      return
+   end
+
+   method assertEqual(expected, actual)
+      if expected ~== actual then {
```

Comment:
Is this supposed to test strings only?

Reply:
Looking at use cases, this should be `expected ~=== actual ` I think.

Reply:
I tried ~= that required two integers but ~== did fine for now. It would convert the numbers before the conversion into strings anyway. Not sure what the advantage of ~=== would be? Anyone could give me an example?

Reply:
`===` supports all types and nit type specific like `=` and `==`.  It works with &null, tables and lists for example. Notice that with structures the check is just for the reference and not actual contents. So
```
L1 := L2 := [ ]
L3 := [ ]
 ```
Ao in the above you will get `L1 === L2`, but `L1~===L3`

Reply:
The problem I see is that `===` prevents data conversion. `1 == "1"` will be valid but `1 === "1"` will fail. Here `L1 == L2´ will cause a runtime error but `L1 === L2´ will succeed. This is the reason I came up with assert { expression } based on t.verify { }. I also noticed that an assertion with parameters that fail won't even enter the assertion method, hence assert and verify use {}. If you have a solution for all those problems, I'm listening!

Reply:
Why do you want 1 to equal "1"?  for a test framework I think you have to be a bit more explicit about type conversion. Because by that logic all of the following would  still fail:
```
 "1" == "1.0"
   1 == "1.0"
   1 == "01"
```
If I want to allow mixing types, I'd do something like:
```
expected ~= actual & expected ~== actual & expected ~=== actual
```
That would allow all possible type coercions to happen in the "correct" order, and cover the cases I listed above and also the structures too.

Reply:
My initial idea was to make assertEqual as convenient as possible and fully benefiting from implicit type conversion. You have however gave me to ponder about whether this should be implemented or not. I further investigated your solution and it would crash if `expected` is of type string. The edge cases would have to be handled. Perhaps ensuring matching types is the programmer's responsibility.

Reply:
my bad. `=` and `==` would cause a runtime error if you supply the wrong time operand. To guard against that I'd try:
```
integer(expected) ~= integer(actual) & string(expected) ~== string(actual) & expected ~=== actual
```

Reply:
And now that might be the wrong logic!